mlflow/pytorch/_pytorch_autolog.py

```diff
@@ -55,264 +52,297 @@ def _get_optimizer_name(optimizer):
 
-    def fit(original, self, *args, **kwargs):
+            if hasattr(optimizer, "defaults"):
+                self.client.log_params(self.run_id, optimizer.defaults)
+
+    def on_train_end(self, trainer, pl_module):
         """
-        Patching trainer.fit method to add autolog class into callback
+        Logs the model checkpoint into mlflow - models folder on the training end
+
+        :param trainer: pytorch lightning trainer instance
+        :param pl_module: pytorch lightning base module
+        """
+        # manually flush any remaining metadata from training
+        self.metrics_logger.flush()
+        self.client.flush(synchronous=True)
+
+    def on_test_end(self, trainer, pl_module):
+        """
+        Logs accuracy and other relevant metrics on the testing end
+
+        :param trainer: pytorch lightning trainer instance
+        :param pl_module: pytorch lightning base module
+        """
+        self.client.set_tags(self.run_id, {"Mode": "testing"})
+        self.client.flush(synchronous=True)
+
+        self.metrics_logger.record_metrics({
+            key: float(value)
+            for key, value in trainer.callback_metrics.items()
+        })
+        self.metrics_logger.flush()
+
+    def _early_stop_check(self, early_stop_callback):
         """
-        return _run_and_log_function(self, original, args, kwargs)
+        Logs all early stopping metrics
 
-    return fit
+        :param early_stop_callback: Early stopping callback object
+        """
+        if early_stop_callback.stopped_epoch != 0:
+            if hasattr(early_stop_callback, "stopped_epoch"):
+                self.metrics_logger.record_metrics(
+                    {"stopped_epoch": early_stop_callback.stopped_epoch}
+                )
+                restored_epoch = early_stop_callback.stopped_epoch - max(
+                    1, early_stop_callback.patience
+                )
+                self.metrics_logger.record_metrics({"restored_epoch": restored_epoch})
+
+            if hasattr(early_stop_callback, "best_score"):
+                self.metrics_logger.record_metrics(
+                    {"best_score": float(early_stop_callback.best_score)}
+                )
+
+            if hasattr(early_stop_callback, "wait_count"):
+                self.metrics_logger.record_metrics(
+                    {"wait_count": early_stop_callback.wait_count}
+                )
+
+
+def _log_early_stop_params(early_stop_callback, client, run_id):
+    """
+    Logs Early Stop parameters into mlflow
+
+    :param early_stop_callback: Early stopping callback dict
+    """
+    if hasattr(early_stop_callback, "monitor"):
+        client.log_params(
+            run_id,
+            {"monitor": early_stop_callback.monitor},
+        )
+
+    if hasattr(early_stop_callback, "mode"):
+        client.log_params(
+            run_id,
+            {"mode": early_stop_callback.mode},
+        )
```

Comment:
@dbczumar Question: is the following code more efficient than the existing code?

```suggestion
    params = {}
    if hasattr(early_stop_callback, "monitor"):
        params["monitor"] = early_stop_callback.monitor

    if hasattr(early_stop_callback, "mode"):
        params["monitor"] = early_stop_callback.mode
   
    client.log_params(run_id, params)
```